### DISCUSSION TITLE
Postgres: FLOAT column gets casted to SERIAL when being primary key

### DISCUSSION THREAD

**Discussion by bgiebl** [sqlalchemy/sqlalchemy#11810] [Usage Questions] [answered]
## Summary
sqlalchemy compiles a primary key FLOAT column to SERIAL by using a postgres database.
This behaviour happens only when there are no foreign key relationships on the column defined and there is no database default defined.
I think this behaviour is a bug and i can propose a fix for it, but i am posting it here to be sure this behaviour is not intended.

## Recreating the problem

```
from sqlalchemy import create_engine, Table, Column, String, MetaData, Float
from sqlalchemy.schema import CreateTable

engine = create_engine("postgresql://scott:tiger@127.0.0.1:5432/test")
metadata = MetaData()

user_table = Table(
        'user', metadata,
        Column('id', Float, primary_key=True),
        Column('name', String)
        )

sql = str(CreateTable(user_table).compile(engine))
print(sql)
```
### Output
```
CREATE TABLE "user" (
        id SERIAL NOT NULL,
        name VARCHAR,
        PRIMARY KEY (id)
)
```
### Expected Behaviour
```
CREATE TABLE "user" (
        id FLOAT NOT NULL,
        name VARCHAR,
        PRIMARY KEY (id)
)
```
## Proposed fix
The problem seems to happen with postgres since in this line
https://github.com/sqlalchemy/sqlalchemy/blob/22cbc7dcb48c946dda66704797665289965eb22e/lib/sqlalchemy/dialects/postgresql/base.py#L2164
there is checked if the primary key column is also an autoincrement column.
So i think it is not a problem by postgres but rather by `_autoincrement_column`.
In the documentation https://docs.sqlalchemy.org/en/20/core/metadata.html#sqlalchemy.schema.Column.params.autoincrement there is written 

> Set up “auto increment” semantics for an integer primary key column with no foreign key dependencies

which indicates that this should only happen with INTEGER and not FLOAT datatypes.
So why does a FLOAT column getting interpreted as an autoincrement column?
It can be traced back to `PrimaryKeyConstraint._autoincrement_column`
https://github.com/sqlalchemy/sqlalchemy/blob/22cbc7dcb48c946dda66704797665289965eb22e/lib/sqlalchemy/sql/schema.py#L5092-L5097
where it is checked if the data type of the column has no type affinity to INTERGERTYPE and  **NUMERICTYPE**.
FLOAT has a type affinity to NUMERICTYPE so the whole check equals to False.
I know that there are some SQL dialects which support autoincrement on FLOAT columns but this is not the case for Postgres.
So depending on the source of the problem there are two possible fixtures:

1. As defined in the documentation, autoincrement can only be applied to INTEGER columns:
    The fix would be to delete line 5096 in schema.py.
2. Autoincrement can be applied to FLOAT columns in some dialects but not in Postgres:
    The fix would be to add an additional type check in `PGDDLCompiler.get_column_specification`
    for a type affinity of the column to INTEGERTYPE.

If you agree that this is unintended behaviour and a bug, i can provide a pull request with one of the fixes.

**Comment by CaselIT** [comment, marked as answer]
Hi,

Definitely seems like a bug. strange that it was never reported before.
I've opened https://github.com/sqlalchemy/sqlalchemy/issues/11811 so let's continue here.
